### ISSUE TITLE
[CSS] OPCPublisher marketplace module: Error 500 Executing command for operation ["create"] failed.

### ISSUE THREAD

**Issue by kyichii** [Azure/Industrial-IoT#1425]
As requested by the IIoT PM for the OPCPublisher Module, I am creating this issue on behalf of a customer that is encountering the following error with the Market Place OPC Publisher module.

Currently, the customer cannot get the module to be started. After some initial troubleshooting where we resolved a DNS configuration issue we end up with the following error in the logs.

```
HTTP Response: 

Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2020_07_07.GeneratedCode.SwaggerException`1[Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2020_07_07.GeneratedCode.ErrorResponse]: Error
 
Status: 500
Response:
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2020_07_07.GeneratedCode.EdgeletHttpClient.CreateModuleAsync(String api_version, ModuleSpec module, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Edge.Util.TaskEx.TimeoutAfter[T](Task`1 task, TimeSpan timeout) in /home/vsts/work/1/s/edge-util/src/Microsoft.Azure.Devices.Edge.Util/TaskEx.cs:line 128
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/versioning/ModuleManagementHttpClientVersioned.cs:line 155
<3> 2021-10-08 04:22:30.395 +00:00 [ERR] [Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory] - Executing command for operation ["create"] failed.
```

Please see the following support bundle for more information.
[support_bundle_verbose.zip](https://github.com/Azure/Industrial-IoT/files/7325890/support_bundle_verbose.zip)

**Comment by daelsala** [comment]
@cristipogacean any updates on this?

**Comment by cristipogacean** [comment]
@kyichii, the exception generated by the edgelet is rather generic and is not providing a clear root for this failure :
`Error calling Create module OPCPublisher: Could not create module OPCPublisher caused by: Could not get module OPCPublisher`

One possible cause of the opc publisher module start failure is that the "binds" in the "Container Create Options" setting fails either due to lack of access to the host's file system of or missing the source folder. Can you try to start the module without this `"Binds":["/iiotedge:/appdata"]` setting just for a quick test? if this works then make sure the /iiotedge exists and is accessible by docker runtime.
The other (unlikely) possible cause is the presence of `"Dns":["49.205.72.130"]`. you could temporarily remove it just to see if this is causing the issue. 

Can you also check if the image of the publisher was properly pulled locally from the mcr? in the logs I see the image gets pulled over and over again.

**Comment by kyichii** [comment]
@cristipogacean Thank you for the follow-up. I have asked the customer to make the changes that you requested and it appears that we are still receiving the same `Error calling Create module OPCPublisher: Could not create module OPCPublisher caused by: Could not get module OPCPublisher` error.

I can see that in the logs, it appears to successfully pull the image, but it cannot connect to it.

> Oct 13 19:14:49 shreesha-VirtualBox iotedged[30699]: 2021-10-13T13:44:49Z [INFO] - Successfully pulled image mcr.microsoft.com/iotedge/opc-publisher:latest
> Oct 13 19:14:49 shreesha-VirtualBox iotedged[30699]: 2021-10-13T13:44:49Z [INFO] - Creating module OPCPublisher...
> Oct 13 19:14:49 shreesha-VirtualBox iotedged[30699]: 2021-10-13T13:44:49Z [INFO] - Starting new listener for module OPCPublisher
> Oct 13 19:14:49 shreesha-VirtualBox iotedged[30699]: 2021-10-13T13:44:49Z [ERR!] - Internal server error: Could not create module OPCPublisher
> Oct 13 19:14:49 shreesha-VirtualBox iotedged[30699]: 	caused by: Could not get module OPCPublisher

I have included the latest support bundle here for the full context.
[support_bundle_binds_removed.zip](https://github.com/Azure/Industrial-IoT/files/7342215/support_bundle_binds_removed.zip)

 
In addition, the customer was able to confirm that the image was pulled successfully and they were able to run it standalone as per the below logs
 
> shreesha@shreesha-VirtualBox:~$ sudo docker images 
> [sudo] password for shreesha:
> REPOSITORY                                   TAG       IMAGE ID       CREATED        SIZE
> mcr.microsoft.com/azureiotedge-diagnostics   1.1.6     b2d4504bbe94   4 weeks ago    98MB
> mcr.microsoft.com/azureiotedge-hub           1.1       fc2272feca50   6 weeks ago    173MB
> mcr.microsoft.com/azureiotedge-agent         1.1       6546d51aa2b8   6 weeks ago    175MB
> dockerhubcommon/vander_modbus                latest    8dec63c9b268   8 weeks ago    243MB
> mcr.microsoft.com/iotedge/opc-publisher      latest    fa74e9aff78d   3 months ago   114MB
> 
> shreesha@shreesha-VirtualBox:~$ sudo docker run -it --mount type=bind,src=/iiotedge,dst=/appdata --env EdgeHubConnectionString="HostName=ESMIOTHUB.azure-devices.net;DeviceId=ESMDEVICE1;SharedAccessKey=hjvJksbpZNTB23UbmBkSSF/mOPzwmIeDV/owDLXZyD0=" fa74e9aff78d --pf /appdata/publishednodes.json -aa --op 10000 
> [13:35:24 INF Root] Starting module OpcPublisher version 2.8.0.1. 
> [13:35:24 INF Root] Initiating prometheus at port 9702/metrics 
> [13:35:25 INF Root] Prometheus metric server started. 
> opcstacktracemask set to: 0x0
> [13:35:27 INF Microsoft.Azure.IIoT.Module.Framework.Client.IoTSdkFactory] 0: Device ESMDEVICE1 reconnected due to Connection_Ok. 
> [13:35:27 INF Microsoft.Azure.IIoT.Module.Framework.Hosting.ModuleHost] Initialize device twin for ESMDEVICE1 - standalone 
> [13:35:28 INF Microsoft.Azure.IIoT.Module.Framework.Hosting.ModuleHost] Applying initial desired state. 
> [13:35:28 INF Microsoft.Azure.IIoT.Module.Framework.Hosting.ModuleHost] Reporting currently initial state. 
> [13:35:29 INF Microsoft.Azure.IIoT.Module.Framework.Hosting.ModuleHost] Module Host started. 
> [13:35:29 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Engine.LegacyJobOrchestrator] File /appdata/publishednodes.json has changed, last known hash null, new hash 39D30F2D3F4DE5A4F57D00CF915DBBF97F62EB6C808DCD0B32CCB05C4B7973CA, reloading... 
> [13:35:29 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Read 1 items from published nodes file in 00:00:00.1192948 
> [13:35:29 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Models.PublishedNodesJobConverter] Converted items to jobs in 00:00:00.1119054 
> [13:35:29 INF Microsoft.Azure.IIoT.OpcUa.Edge.Publisher.Engine.LegacyJobOrchestrator] Job opc.tcp://192.168.0.104:53530/OPCUA/SimulationServer_93a66fa7c0c7b186b205cb250b1848962c6e478c loaded. DataSetWriters endpoints: opc.tcp://192.168.0.104:53530/OPCUA/SimulationServer 
> [13:35:30 INF Microsoft.Azure.IIoT.Agent.Framework.Agent.WorkerSupervisor] Creating new worker with id StandalonePublisher_0 
> [13:35:30 INF Microsoft.Azure.IIoT.Agent.Framework.Agent.WorkerSupervisor] Starting worker 'StandalonePublisher_0'... 
> [13:35:30 INF Microsoft.Azure.IIoT.Agent.Framework.Agent.Worker] Worker StandalonePublisher_0: {} 
> [13:35:30 INF Microsoft.Azure.IIoT.Agent.Framework.Agent.Worker] Worker: StandalonePublisher_0 processing job: opc.tcp://192.168.0.104:53530/OPCUA/SimulationServer_93a66fa7c0c7b186b205cb250b1848962c6e478c, mode: Active 
> [13:35:31 INF Microsoft.Azure.IIoT.OpcUa.Protocol.Services.DefaultSessionManager] Subscription 'opc.tcp://192.168.0.104:53530/OPCUA/SimulationServer_4ca1ffa717be657d4ffa481d9efb66b84b51e592' registered/updated in session '93a66fa7c0c7b186b205cb250b1848962c6e478c' in state Init 
> [13:35:31 INF Microsoft.Azure.IIoT.OpcUa.Protocol.Services.DefaultSessionManager] Subscription 'opc.tcp://192.168.0.104:53530/OPCUA/SimulationServer_4ca1ffa717be657d4ffa481d9efb66b84b51e592' registered/updated in session '93a66fa7c0c7b186b205cb250b1848962c6e478c' in state Init
>

**Comment by kyichii** [comment]
@cristipogacean Any additional insight into this matter?

**Comment by cristipogacean** [comment]
Hi @kyichii, 

I retested the same scenario in our environment and cannot replicate this issue. The configuration of the OPC publisher seem to be fine and normally it should work without issues. As I mentioned previously the exceptions logged in the traces of edgeAgent respectively iotedge daemon are rather generic and do not provide a clear information on the cause.

Doing further investigations in the latest support bundle I suspect the setup of the iotedge runtime on the host is causing this issue so my suggestion would be to test the OPC publisher on a freshly installed iotedge runtime and make sure the docker engine in use is moby and not a different engine.  Here are my findings that support my suggestion:

In check.json logfile I noticed the followings:
```
   "additional_info": {
        "docker_version": "20.10.9+azure-1",
        "iotedged_version": "1.1.6",
```
and 
```
        "container-engine-is-moby": {
            "result": {
                "result": "warning",
                "details": [
                    "Device is not using a production-supported container engine (moby-engine).\nPlease see https://aka.ms/iotedge-prod-checklist-moby for details."
                ]
            },
            "additional_info": {
                "docker_server_version": "20.10.9+azure-1",
                "moby_runtime_uri": null
            }
        },
 ```
With this I'm assuming the docker engine on the host machine is not a production-supported version so it might be an strange incompatibility that prevents the OPC publisher module to start. 

Furthermore I see the iotedge runtime version is 1.1.6, but the edgeAgent ant edgeHub that are actually running report version 1.1.5. Probably this is expected since it looks like 1.1.6 tag only contains iotedge daemon binaries :
edgeAgent:
``` 
<6> 2021-10-13 13:40:52.187 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Agent.Service.Program] - Version - 1.1.5.46127017 (7681cae0e56a036f9659814ad559237f88d96fa3)
```
edgeHub:
```
<6> 2021-10-13 13:41:30.660 +00:00 [INF] [EdgeHub] - Version - 1.1.5.46127017 (7681cae0e56a036f9659814ad559237f88d96fa3)
```

Looking in the IotEdge repo I noticed some bugfixed related to the container socket: https://github.com/Azure/azure-iotedge/releases/tag/1.1.7 and https://github.com/Azure/azure-iotedge/releases/tag/1.1.6. I'm not sure if this is related or not to your issue, but I would suggest to uninstall the iotedge on the host, make sure any docker engine instance is removed and then install intedge 1.1.7.

If this still not works I would suggest to try to start just the the OPC publisher module alone on the edge device- without the modbus module. 

I hope this helps us moving forward.
